Capstone 3 - Data Wrangling

In [3]:
#Import pandas, matplotlib.pyplot, and seaborn 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
!pip install openpyxl

In [5]:
### Load Dataset

file = '/Users/elliotgaynon/Desktop/CAP3/Capstone3/data/raw/Online Retail.xlsx'

data_raw = pd.read_excel(file)

In [6]:
### Confirm DataFrame is loaded
data_raw

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

**I. Verify Columns and Data Types**

In [7]:
data_raw.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [8]:
data_raw.shape

(541909, 8)

We have 541909 rows and 8 columns

In [9]:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


Categorical variables appear as objects, continuous variables appear as either integers or floats, and my InvoiceDate appears as a datetime object. At some point downstream will have to turn CustomerID from a float to an integer.  However, this looks good.  Now I will check for missing values.

***II. Check for Missing Values + Data Cleaning***

In [10]:
### Checking DataFrame for missing values

data_raw.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

***I have 1,454 missing Description values and 135,080 missing values for CustomerID.***


For missing Description values, I will leave them as is since they won't be used for RFM features or my target consrtuction. Dropping these rows would gain me nothing and filling them would be unnecessary.

On the other hand, CustomerID is what I'll be aggregating and computing RFM metrics with.  If I have no data here, I won't be able to perform my analysis.

In [11]:
### Dropping Missing CustomerID

data_raw = data_raw.dropna(subset = ['CustomerID'])


In [12]:
###Confirm missing values are gone in CustomerID
data_raw.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

It looks like when I got rid of missing CustomerID values, my missing description values appeared to disappear as well.  I want to see if any missing descriptions existed only among rows with valid CustomerIDs

In [13]:
data_raw[data_raw['CustomerID'].notna()]['Description'].isna().sum()

np.int64(0)

Zero descriptions existed among valid CustomerIDs. This confirms that all missing descriptions were tied to missing CustomerIDs.  Thus, no usable customer data was lost. Now all rows with missing customerID data, as well as missing product descriptions, are gone.

***III. Checking for canceled invoices, negative or zero prices and Negative quantities***

The online retail dataset website states "If the Invoice code starts with letter 'c', it indicates a cancellation."  A cancelled invoice means there's no revenue.  Thus, in order to calculate my target variable, Monetary Value, and Frequency, I must remove these invoices.

In [14]:
### Inspect InvoiceNo

num_canceled = data_raw['InvoiceNo'].astype(str).str.startswith('C').sum()
num_canceled

np.int64(8905)

There are 8905 Invoices that start with the letter 'C'.  They must be removed.

Since Revenue = Quantity x UnitPrice negative quantities of items or unit prices must be removed.  We must only have revenue generating sales in the model of customer value.  Thus, these items must be removed as well since we are trying to predict future sales.

In [15]:
### Check for Non-Positive Quantities

num_neg_qty =(data_raw['Quantity'] <= 0).sum()
num_neg_qty

np.int64(8905)

There are 8905 rows who's quantities are less than or equal to zero.  This is the same number as our canceled orders!

In [16]:
### Check for Non-Positive Prices

num_neg_prc = (data_raw['UnitPrice'] <= 0).sum()
num_neg_prc

np.int64(40)

There are 40 rows where the price is less than or equal to zero. 

In [17]:
### Overlap Check

data_raw[
    data_raw['InvoiceNo'].astype(str).str.startswith('C') |
    (data_raw['Quantity'] <= 0) |
    (data_raw['UnitPrice'] <= 0)
].shape[0]

8945

8945 rows must be removed in total

In [18]:
### Revenue Sanity check

(data_raw['Quantity'] * data_raw['UnitPrice']).describe()

count    406829.000000
mean         20.401854
std         427.591718
min     -168469.600000
25%           4.200000
50%          11.100000
75%          19.500000
max      168469.600000
dtype: float64

My minimum value is a very large negative value. This confirms quantity and unit price must be cleaned out.

***IV. Removing canceled invoices, non-positive quantity/price***

In [19]:
### remove canceled invoices

before = data_raw.shape[0]

mask_canceled = data_raw['InvoiceNo'].astype(str).str.startswith('C')
data_raw = data_raw[~mask_canceled]

after = data_raw.shape[0]
print(f"Dropped {before - after} canceled invoice rows")
print(f"Remaining rows: {after}")

Dropped 8905 canceled invoice rows
Remaining rows: 397924


In [20]:
### Remove Non-Positive Quantities

before = data_raw.shape[0]

data_raw = data_raw[data_raw['Quantity'] > 0]

after = data_raw.shape[0]
print(f"Dropped {before - after} rows with non-positive Quantity")
print(f"Remaining rows: {after}")

Dropped 0 rows with non-positive Quantity
Remaining rows: 397924


This makes sense since non-positive quantities was the same as canceled transactions.

In [21]:
### Remove Non-Positive Pricing

before = data_raw.shape[0]

data_raw = data_raw[data_raw['UnitPrice'] > 0]

after = data_raw.shape[0]
print(f"Dropped {before - after} rows with non-positive Quantity")
print(f"Remaining rows: {after}")

Dropped 40 rows with non-positive Quantity
Remaining rows: 397884


In [22]:
### Final Confirmation
print(f"Final cleaned rows: {data_raw.shape[0]}")

Final cleaned rows: 397884


In [23]:
### Checking for any missing values

data_raw.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

No Missing values.  Can move on.

***V. Creation of Revenue Column***

Before moving on, I want to double confirm UnitPrice and Quantity are greater than 0.

In [24]:
### Sanity Check

(data_raw['Quantity'] <= 0).sum(), (data_raw['UnitPrice']<=0).sum()

(np.int64(0), np.int64(0))

All values of quantity and unit price are where we want them.  

In [25]:
### Create the Revenue Column - Revenue = Quantity x UnitPrice

data_raw['Revenue'] = data_raw['Quantity'] * data_raw['UnitPrice']

In [26]:
### Description of new Revenue Column

data_raw['Revenue'].describe()

count    397884.000000
mean         22.397000
std         309.071041
min           0.001000
25%           4.680000
50%          11.800000
75%          19.800000
max      168469.600000
Name: Revenue, dtype: float64

Minimum is greater than 0.  No huge negatives. 

In [27]:
### Quick Spot check 

data_raw[['Quantity', 'UnitPrice', 'Revenue']].head()

Quantity  UnitPrice  Revenue
0         6       2.55    15.30
1         6       3.39    20.34
2         8       2.75    22.00
3         6       3.39    20.34
4         6       3.39    20.34

***VI. Data Handling and Cutoff***

For selecting the cutoff date, I must consider a minimum window of past transactions in order to calculate RFM(Recency, Frequency, and MonetaryValue).  Thus, cutoff date + 60 days must fit within the data set.  Also I want to choose a cutoff that isn't too close to the least transaction date.

In [28]:
### Inspect Dataset Dates - confirm datetime

data_raw['InvoiceDate'].info()

<class 'pandas.core.series.Series'>
Index: 397884 entries, 0 to 541908
Series name: InvoiceDate
Non-Null Count   Dtype         
--------------   -----         
397884 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 6.1 MB


In [29]:
### check first and last transactions

print("Earliest transaction:", data_raw['InvoiceDate'].min())
print("Latest transaction:", data_raw['InvoiceDate'].max())

Earliest transaction: 2010-12-01 08:26:00
Latest transaction: 2011-12-09 12:50:00


The transactions occured from December 1st, 2010 until December 9th, 2011.  A little over a year.

The latest safe cutoff = the latest transaction minus 60 days.

In [30]:
### Choose Cutoff Date

import datetime

last_date = data_raw['InvoiceDate'].max()

cutoff_date = last_date - datetime.timedelta(days=60)

print("Last Date:", last_date)
print("Cutoff date:" , cutoff_date)

Last Date: 2011-12-09 12:50:00
Cutoff date: 2011-10-10 12:50:00


In [31]:
### Check Transaction Volume Around cutoff date 

transactions_per_day = data_raw.groupby('InvoiceDate').size()
transactions_per_day[cutoff_date - pd.Timedelta(days=10) : cutoff_date + pd.Timedelta(days=10)]

InvoiceDate
2011-09-30 12:58:00    14
2011-09-30 13:06:00    17
2011-09-30 13:09:00    33
2011-09-30 13:15:00    23
2011-09-30 13:21:00     5
                       ..
2011-10-20 12:31:00    11
2011-10-20 12:34:00    40
2011-10-20 12:40:00     1
2011-10-20 12:41:00     6
2011-10-20 12:43:00    95
Length: 1157, dtype: int64

Cutoff date appears to fall in a reasonably active period.  No major holidays within this range.
Thus, 

Features = all transactions ≤ 2011-10-10

Target = sum of revenue for each customer 2011-10-11 → 2011-12-09 (60-day window)

In [32]:
### Convert CustomerID from float to int

data_raw['CustomerID'] = data_raw['CustomerID'].astype(int)

In [33]:
### Sanity Check

data_raw['CustomerID'].dtype

dtype('int64')

In [34]:
### saving cleaned data_set as csv

data_raw.to_csv("cleaned_transactions.csv", index=False)

***VII. Filter Pre-Cutoff Data***

I am filtering transactions before the cutoff date (2011-10-10) in order to see what each customer looked like before the cutoff and then see how much revenue they generated over a 60 day period.

In [35]:
### Creation of Feature Window

target_window_days = 60
target_end_date = cutoff_date + pd.Timedelta(days=target_window_days)

cutoff_date, target_end_date

(Timestamp('2011-10-10 12:50:00'), Timestamp('2011-12-09 12:50:00'))

A cutoff timestamp of October 10, 2011 was selected, leaving a 60-day forward-looking window through December 9, 2011 to define the target variable.

In [37]:
### Splitting Transactions before (features) and after(target) Cutoff Date

features_txn = data_raw[data_raw['InvoiceDate'] <= cutoff_date].copy()
target_txn = data_raw[
    (data_raw['InvoiceDate'] > cutoff_date) &
    (data_raw['InvoiceDate'] <= target_end_date)
].copy()

In [38]:
### Sanity Check

print("Feature window date range:")
print(features_txn['InvoiceDate'].min(), "to", features_txn['InvoiceDate'].max())

print("\nTarget window date range:")
print(target_txn['InvoiceDate'].min(), "to", target_txn['InvoiceDate'].max())

Feature window date range:
2010-12-01 08:26:00 to 2011-10-10 12:50:00

Target window date range:
2011-10-10 12:51:00 to 2011-12-09 12:50:00


No feature window dates occur past the cutoff date.  No target window dates occur past the target end date. In other words, features are comprised ONLY of past data.  Targets are ONLY future looking.

***VIII. Customer Level Aggregation via Feature Transactions***

In [42]:
features_txn.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Revenue  
0 2010-12-01 08:26:00       2.55       17850  United Kingdom    15.30  
1 2010-12-01 08:26:00       3.39       17850  United Kingdom    20.34  
2 2010-12-01 08:26:00       2.75       17850  United Kingdom    22.00  
3 2010-12-01 08:26:00       3.39       17850  United Kingdom    20.34  
4 2010-12-01 08:26:00       3.39       17850  United Kingdom    20.34

The dataset is currently structured at the transaction level, with one row per purchase. Because the business question focuses on identifying which customers should be prioritized for promotions, the data must be transformed to one row per customer. This aggregation enables the computation of Recency, Frequency, and Monetary (RFM) metrics, which summarize customer-level purchasing behavior.

In [43]:
### Calculate Recency - How recently was this customer active at the cutoff?

last_purchase = (features_txn.groupby('CustomerID')['InvoiceDate'].max())

recency = ((cutoff_date - last_purchase).dt.days.rename('recency'))

In [44]:
### Calculate Frequency - How often does customer purchase from company?

frequency = (features_txn.groupby('CustomerID')['InvoiceNo'].nunique().rename('frequency'))

In [45]:
### Calculate MonetaryValue - How valuable is the customer historically?

monetary = (features_txn.groupby('CustomerID')['Revenue'].sum().rename('monetary'))

***IX. Creation of Proxy Feature - AOV***

Since there is no column for "promotional response", a proxy feature can be calculated to approximate likelihood to respond. I will calculate AOV (Average Order Value), a proxy for spending habits. AOV tells how much a customer spends each time they buy. So a customer with a high AOV spends more per purchase.  Theoretically, if a promotion can target this customer, they will generate more revenue per response.

In [48]:
### Calculate AOV - Total money spent ÷ number of invoices

aov = (monetary / frequency).rename('avg_order_value')

***X. Creation of Features Table (RFM + Proxy)***

In [50]:
### Create table
features = pd.concat([recency, frequency, monetary, aov],axis=1).reset_index()

In [51]:
### Confirm table

features.head()

CustomerID  recency  frequency  monetary  avg_order_value
0       12346      265          1  77183.60     77183.600000
1       12347       69          5   2790.86       558.172000
2       12348       14          4   1797.24       449.310000
3       12350      249          1    334.40       334.400000
4       12352       11          7   2194.31       313.472857

In [52]:
### Any NaN values and how many?

features.isna().sum()

CustomerID         0
recency            0
frequency          0
monetary           0
avg_order_value    0
dtype: int64

Nope.

***XI. Target Variable Construction*** 

This is the other half of the time split. The prediction horizon is anything that happened between the cutoff date and sixty days into the future.

In [55]:
### Aggregate total Customer Revenue within the 60 day window.

target = (target_txn.groupby('CustomerID')['Revenue'].sum().rename('target_60day_revenue').reset_index())

***XII. Merge Features and Target Tables to create DataFrame***

In [56]:
df = features.merge(target,on='CustomerID',how='left')

In [58]:
### Assign 0 to to NaN purchases

df['target_60day_revenue'] = df['target_60day_revenue'].fillna(0)

Customers who didn't buy anything still matter.  Filling NaN with 0 avoids selection bias because they are legitimate observations depicting customers with 0 short-term value.

***XIII. Sanity Checks and verifications***

In [60]:
### Statistical overview of my DataFrame
df.describe()

CustomerID      recency    frequency       monetary  avg_order_value  \
count   3731.000000  3731.000000  3731.000000    3731.000000      3731.000000   
mean   15276.365854    92.700884     3.676494    1748.219787       408.856860   
std     1720.658779    88.850967     6.271227    7296.654491      1357.896516   
min    12346.000000     0.000000     1.000000       2.900000         2.900000   
25%    13779.000000    18.000000     1.000000     289.355000       177.397500   
50%    15254.000000    60.000000     2.000000     602.840000       296.640000   
75%    16754.500000   150.500000     4.000000    1457.835000       434.937222   
max    18287.000000   313.000000   145.000000  203948.190000     77183.600000   

       target_60day_revenue  
count           3731.000000  
mean             550.139268  
std             3615.879271  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%              455.100000  
max           168469.600000

Will have to explore this more in EDA.

In [62]:
### Confirming recency and tagret variable is not negative numbers
(df['recency'] < 0).sum()
(df['target_60day_revenue'] < 0).sum()

np.int64(0)

Both are zero. This is exactly what I want.

In [64]:
### Final confirmation
df.head()

CustomerID  recency  frequency  monetary  avg_order_value  \
0       12346      265          1  77183.60     77183.600000   
1       12347       69          5   2790.86       558.172000   
2       12348       14          4   1797.24       449.310000   
3       12350      249          1    334.40       334.400000   
4       12352       11          7   2194.31       313.472857   

   target_60day_revenue  
0                  0.00  
1               1519.14  
2                  0.00  
3                  0.00  
4                311.73

From the DataFrame, one can see a table where we have our target variable "target_60day_revenue" and our RFM + proxy features.  Their relationships will be explored further in EDA.

***XIV. Saving DataFrame for EDA***

In [65]:
df.to_csv("capstone3_customer_modeling_table_For_EDA.csv", index=False)